<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>then方法</title>
</head>
<body>
<script>
    const getJson = function (url) {
        return new Promise((resolve, reject) => {
            const xhr = new XMLHttpRequest();
            xhr.open('GET', url);
            xhr.onreadystatechange = handler;
            xhr.responseType = 'json';
            xhr.setRequestHeader('Accept', 'application/json');
            xhr.send();

            function handler() {
                if (this.readyState === 4) {
                    if (this.status === 200) {
                        resolve(this.response.HeWeather6);
                    } else {
                        reject(new Error(this.statusText));
                    }
                }
            }
        })
    }

    // then方法
    // then 第一个参数是 relove 回调函数，第二个参数是可选地，是 reject 回调函数
    // then 返回一个新的 promise 实例，可以采用链式编程
    // getJson('https://free-ap.heweather.net/s6/weather/now?location=beijing&key=8570f6b17578432d86eabe3290e5c659')
    //     .then(data=>{
    //         console.log(data); // 第一个  Array(1)
    //     }).then(data=>{
    //         console.log(data); // undefined
    //     });

    // getJson('https://free-api.heweather.net/s6/weather/now?location=beijing&key=8570f6b17578432d86eabe3290e5c659')
    //     .then(data=>{
    //         console.log("第一个 ",data); // 第一个  Array(1)
    //         return data; // 这里返回的值，在下一个then的参数中接收
    //     }).then(data=>{
    //         console.log("第二个 ",data); // 第二个  Array(1)
    //     });

    // then... catch...
    // Attaches a callback for only the rejection of the Promise.

    // getJson('https://free-ap.heweather.net/s6/weather/now?location=beijing&key=8570f6b17578432d86eabe3290e5c659')
    //     .then(data => {
    //         console.log(data);
    //     }).then(null, err => {
    //     // 错误： Error at XMLHttpRequest.handler
    //     console.log("错误：", err);
    // });

    // 可以改写成：
    // getJson('https://free-api.heweather.net/s6/weather/now?location=beijing&key=8570f6b17578432d86eabe3290e5c659')
    //     .then(data=>{
    //         console.log(data);
    //     }).catch(err=>{
    //     // 错误： Error at XMLHttpRequest.handler
    //     console.log("错误：", err);
    // });

    // 捕捉顺序：依次向下捕捉，捕捉到就暂停，就开始执行catch
    new Promise((resolve, reject) => {
        resolve('p1 -- 成功');
    }).then(res => {
        throw new Error('p2 -- 错误');
    }).then(res => {
        throw new Error('p3 -- 错误');
    }).catch(err => {
        console.log("错误：", err); // 错误： Error: p2 -- 错误
    });

    // 参考理解： https://blog.csdn.net/cxu123321/article/details/106760091
</script>
</body>
</html>